### PR TITLE
Porting iree::WaitHandle to C and adding Windows support.

### LINKED ISSUE #65: iree/base/wait_handle.cc relies on unix specific headers/features
Does not build on Windows.
---
ERROR: C:/src/ireepub/iree/iree/base/BUILD:348:1: Couldn't build file iree/base/_objs/wait_handle_test/wait_handle_test.obj: C++ compilation of rule '//iree/base:wait_handle_test' failed (Exit 1)
iree/base/wait_handle_test.cc(17,10): fatal error: 'unistd.h' file not found
#include <unistd.h>
         ^~~~~~~~~~
1 error generated.
ERROR: C:/src/ireepub/iree/iree/base/BUILD:330:1: Couldn't build file iree/base/_objs/wait_handle/wait_handle.obj: C++ compilation of rule '//iree/base:wait_handle' failed (Exit 1)
iree/base/wait_handle.cc(19,10): fatal error: 'poll.h' file not found
#include <poll.h>
         ^~~~~~~~
1 error generated.

### PR DESCRIPTION
**Pull request by benvanik**
iree_wait_handle_t and iree_wait_set_t allow us to efficiently perform single- and multi-waits on operating system wait primitives (eventfds, HANDLEs, pipes, etc). This is the foundation for our import/export of handles to things like HAL semaphores: we now have a way to shuffle them across the C APIs and mix in- and out-of-process CPU-side handles as well as hardware driver handles in a single place. This makes things like fine-grained cross-process synchronization (such as if running the HAL backend in a sandbox) and heterogenous pipelining (CPU task pool mixing with GPU compute queues) easy peasy.

Currently only win32 and poll/ppoll are supported (as poll is available on all platforms); kqueue (for mac/ios) and epoll (for android/linux/bsd) are significantly better than poll/ppoll though and we'll want to fill those in over time. By splitting out the wait set implementation from the wait primitive implementation we don't end up with a full explosion of code there (an eventfd can be passed to poll, ppoll, or epoll, etc), and we can use the much lighter-weight implementations on platforms that have them available. For example our iree_wait_set_t becomes a reinterpreted epoll/kqueue fd and avoids user-space duplication of data entirely, whereas poll/win32 requires us to emulate what epoll/kqueue do with our own data structures.

Fixes #65.

### CODE REVIEW

**Comment by benvanik** [comment]
(failures are from base benvanik-tasks-pre branch - will fix and rebase)

File: iree/base/wait_handle_test.cc [hunk 1]
```diff
@@ -14,538 +14,803 @@
 
 #include "iree/base/wait_handle.h"
 
-#include <unistd.h>
+#include <thread>
 
-#include <string>
-#include <thread>  // NOLINT
-#include <type_traits>
```
Comment (ScottTodd) [inline iree/base/wait_handle_test.cc:21]: ugh GitHub's diff in this file is so bad.. so many arbitrary breaks between old and new code

File: iree/base/wait_handle_test.cc [hunk 2]
```diff
@@ -14,538 +14,803 @@
...
 }
 
-// Tests polling with WaitAny with multiple wait handles.
-TEST(WaitHandleTest, WaitAnyPoll) {
-  ManualResetEvent fence0;
-  WaitHandle wh0 = fence0.OnSet();
-  ManualResetEvent fence1;
-  WaitHandle wh1 = fence1.OnSet();
+// Tests that we can use set/reset and that certain behavior (such as sets
+// without intervening resets) is allowed. Note that this does not wait and is
```
Comment (ScottTodd) [inline iree/base/wait_handle_test.cc:124]: A bit weary about ASAN/MSAN/TSAN and other similar checks in this file. I'd say go with what our open source CI is actually checking, but someone might need to triage some warnings/errors after merge.

**Review by ScottTodd — APPROVED** [review]

File: iree/base/wait_handle_test.cc (hunk 2 above)
Comment (benvanik) [reply iree/base/wait_handle_test.cc:124]: A lot of the patterns in this file originated from google3 where asan/msan/etc was being used. Still possible for there to be issues as the implementation is really different, but that's even more reason that someone should set up a bot :P

File: iree/base/wait_handle_kqueue.c [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  // TODO(benvanik): kqueue support
+}
+
+void iree_wait_set_free(iree_wait_set_t* set) {
+  // TODO(benvanik): close()
+}
+
+iree_status_t iree_wait_set_insert(iree_wait_set_t* set,
+                                   iree_wait_handle_t handle) {
+  // TODO(benvanik): kqueue support
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/base/wait_handle_kqueue.c:44]: These files will all fail to compile if they are activated, which is a better error :)

**Pull request merged**